stdlib/public/runtime/RuntimeInvocationsTracking.cpp

```diff
@@ -0,0 +1,231 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Track invocations of Swift runtime functions. This can be used for performance
+// analysis.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/DenseMap.h"
+#include "RuntimeInvocationsTracking.h"
+#include "swift/Runtime/HeapObject.h"
+
+using namespace swift;
+
+#define SWIFT_RT_FUNCTION_INVOCATION_COUNTER_NAME(RT_FUNCTION)                 \
+  invocationCounter_##RT_FUNCTION
+
+#define FUNCTION_TO_TRACK(RT_FUNCTION)                                         \
+  uint32_t SWIFT_RT_FUNCTION_INVOCATION_COUNTER_NAME(RT_FUNCTION) = 0;
+
+namespace swift {
+
+// Define counters used for tracking the total number of invocations of runtime
+// functions.
+struct RuntimeFunctionCountersState {
+  // Provide one counter per runtime function being tracked.
+#include "RuntimeInvocationsTracking.def"
+};
+
+} // end namespace swift
+
+
+/// If set, global runtime function counters should be tracked.
+static bool UpdatePerObjectRuntimeFunctionCounters = false;
+/// If set, per object runtime function counters should be tracked.
+static bool UpdateGlobalRuntimeFunctionCounters = false;
+/// TODO: Add support for enabling/disabling counters on a per object basis?
+
+/// Global set of counters tracking the total number of runtime invocations.
+static RuntimeFunctionCountersState RuntimeGlobalFunctionCountersState;
+
+/// The object state cache mapping objects to the collected state associated with
+/// them.
+/// TODO: Do we need to make it thread-safe?
+static llvm::DenseMap<HeapObject *, RuntimeFunctionCountersState> RuntimeObjectStateCache;
+
+/// Define names of runtime functions.
+#define FUNCTION_TO_TRACK(RT_FUNCTION) #RT_FUNCTION,
+
+static const char *RuntimeFunctionNames[] {
+  #include "RuntimeInvocationsTracking.def"
+  nullptr
+};
+
+#define RT_FUNCTION_ID(RT_FUNCTION) ID_##RT_FUNCTION
+
+/// Defines names of enum cases for each function being tracked.
+#define FUNCTION_TO_TRACK(RT_FUNCTION) RT_FUNCTION_ID(RT_FUNCTION),
+
+/// Define an enum where each enumerator corresponds to a runtime function being
+/// tracked. Their order is the same as the order of the counters in the
+/// RuntimeObjectState structure.
+enum RuntimeFunctionNamesIDs : uint32_t {
+  #include "RuntimeInvocationsTracking.def"
+  ID_LastRuntimeFunctionName,
+};
+
+/// The global handler to be invoked on runtime function counters updates.
+static RuntimeFunctionCountersUpdateHandler
+    GlobalRuntimeFunctionCountersUpdateHandler;
+
+/// Define offset for each function being tracked.
+#define FUNCTION_TO_TRACK(RT_FUNCTION)                                         \
```

Comment:
Can you move this next to the include of RuntimeInvocationsTracking.def in the function itself.